src/Tinker.jl

```diff
@@ -1,4 +1,268 @@
+                range.y.left+Int(round(csize.y/2)))
+    zoom_percent(z,zr,center)
+end
+
+## Sets up an image in a separate window with the ability to adjust view
+function init_gui(image::AbstractArray; name="Tinker")
+    # set up window
+    win = Window(name, size(image,2), size(image,1));
+    c = canvas(UserUnit);
+    push!(win, c);
+
+    # set up a zoom region
+    zr = Signal(ZoomRegion(image))
+
+    # create view
+    imagesig = map(zr) do r
+        cv = r.currentview
+        view(image, UnitRange{Int}(cv.y), UnitRange{Int}(cv.x))
+    end;
+    
+    # create a view diagram
+    viewdim = map(zr) do r
+        fvx, fvy = r.fullview.x, r.fullview.y # x, y range of full view
+        cvx, cvy = r.currentview.x, r.currentview.y # x, y range of currentview
+        xfull, yfull =
+            (fvx.right-fvx.left),(fvy.right-fvy.left) # width of full view
+        xcurrent, ycurrent =
+            (cvx.right-cvx.left),(cvy.right-cvy.left) # width of current view
+        # scale
+        xsc,ysc = 0.1*(xcurrent/xfull), 0.1*(ycurrent/yfull)
+        # offset
+        x_off,y_off = cvx.left+(0.01*xcurrent),cvy.left+(0.01*ycurrent)
+        # represents full view
+        rect1 = Rectangle(x_off, y_off, xsc*xfull, ysc*yfull)
+        # represents current view
+        rect2 = Rectangle(x_off+(cvx.left*xsc), y_off+(cvy.left*ysc),
+                          xsc*xcurrent, ysc*ycurrent)
+        return [rect1,rect2]
+    end
+
+    # draw
+    redraw = draw(c, imagesig, zr, viewdim) do cnvs, img, r, vd
+        copy!(cnvs, img) # show image on canvas at current zoom level
+        set_coordinates(cnvs, r) # set canvas coordinates to zr
+        ctx = getgc(cnvs)
+        # draw view diagram if zoomed in
+        if r.fullview != r.currentview
+            drawrect(ctx, vd[1], colorant"blue")
+            drawrect(ctx, vd[2], colorant"blue")
+        end
+    end
+
+    # holds x zoom level
+    xzoom = map(zr) do r
+        100*r.fullview.x.right/(r.currentview.x.right-(r.currentview.x.left-1))
+    end
+
+    # hold y zoom level
+    yzoom = map(zr) do r
+        100*r.fullview.y.right/(r.currentview.y.right-(r.currentview.y.left-1))
+    end
+
+    zpercents = [1.0,1.2,1.5,2.0,2.5,3.0,4.0,8.0]
+    global i = 1 # or: const i = Ref(1)
+
+    # performs proportional zoom in
+    function zoomin(center::XY{Int})
+        global i
+        if 1 <= i <= length(zpercents)
+            if i < length(zpercents)
+                i += 1
+                push!(zr, zoom_percent(zpercents[i],value(zr),center))
+            end
+        else
+            index = 1
+            for n in zpercents
+                if n > value(xzoom)
+                    break
+                end
+                index += 1
+            end
+            i = index
+            push!(zr, zoom_percent(zpercents[i],value(zr),center))
+        end
+    end
+
+    function zoomin()
+        global i
+        if 1 <= i <= length(zpercents)
```

Comment:
Seems like most of this logic could be split into a separate function to avoid duplication? True for `zoom_out` too.

De-duplication is important because if you need to update code/fix bugs, it's nice to have just one place to fix.